Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/common/DccJWKVerification.kt

```diff
@@ -0,0 +1,80 @@
+package de.rki.coronawarnapp.dccticketing.core.common
+
+import com.nimbusds.jose.JWSAlgorithm.ES256
+import com.nimbusds.jose.JWSAlgorithm.PS256
+import com.nimbusds.jose.JWSAlgorithm.RS256
+import com.nimbusds.jose.crypto.ECDSAVerifier
+import com.nimbusds.jose.crypto.RSASSAVerifier
+import com.nimbusds.jose.crypto.bc.BouncyCastleProviderSingleton
+import com.nimbusds.jose.util.X509CertUtils
+import com.nimbusds.jwt.SignedJWT
+import de.rki.coronawarnapp.dccticketing.core.common.DccTicketingException.ErrorCode.JWT_VER_ALG_NOT_SUPPORTED
+import de.rki.coronawarnapp.dccticketing.core.common.DccTicketingException.ErrorCode.JWT_VER_NO_JWKS
+import de.rki.coronawarnapp.dccticketing.core.common.DccTicketingException.ErrorCode.JWT_VER_NO_JWK_FOR_KID
+import de.rki.coronawarnapp.dccticketing.core.common.DccTicketingException.ErrorCode.JWT_VER_NO_KID
+import de.rki.coronawarnapp.dccticketing.core.common.DccTicketingException.ErrorCode.JWT_VER_SIG_INVALID
+import de.rki.coronawarnapp.dccticketing.core.transaction.DccJWK
+import okio.ByteString.Companion.decodeBase64
+import org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey
+import timber.log.Timber
+import java.security.PublicKey
+import java.security.interfaces.RSAPublicKey
+import javax.inject.Inject
+
+class DccJWKVerification @Inject constructor() {
+
+    init {
+        X509CertUtils.setProvider(BouncyCastleProviderSingleton.getInstance())
+    }
+
+    fun verify(jwt: String, jwkSet: Set<DccJWK>) {
+        // 1. Check for empty jwkSet
+        if (jwkSet.isEmpty()) throw DccTicketingException(JWT_VER_NO_JWKS)
+
+        // 2. Check alg of JWT
+        val signedJWT = try {
+            SignedJWT.parse(jwt)
+        } catch (e: Exception) {
+            Timber.e("Can't parse JWT token $jwt", e)
+            throw DccTicketingException(JWT_VER_ALG_NOT_SUPPORTED)
+        }
+
+        if (signedJWT.header.algorithm !in listOf(ES256, PS256, RS256))
+            throw DccTicketingException(JWT_VER_ALG_NOT_SUPPORTED)
+
+        // 3. Extract kid from JWT
+        if (signedJWT.header.keyID.isNullOrEmpty()) throw DccTicketingException(JWT_VER_NO_KID)
+
+        // 4. Check for empty jwkSet
+        if (jwkSet.none { it.kid == signedJWT.header.keyID }) throw DccTicketingException(JWT_VER_NO_JWK_FOR_KID)
+
+        // 5. Filter jwkSet by kid
+        jwkSet.filter { it.kid == signedJWT.header.keyID }.forEach {
+            try {
+                // 6. Verify signature
+                val publicKey = X509CertUtils.parse(it.x5c.first().decodeBase64()?.toByteArray()).publicKey
```

Comment:
can `dccJWKConverter.createX509Certificate(jwk = it)`   be used ?

Reply:
`DccJWKConverter` has not been merged yet. I tested the code anyway, but unfortunately, it cannot satisfy the JWT library as it returns the public key in a different format. Eg. exception:

```
java.lang.ClassCastException: class sun.security.ec.ECPublicKeyImpl cannot be cast to class org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey (sun.security.ec.ECPublicKeyImpl is in module jdk.crypto.ec of loader 'platform'; org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey is in unnamed module of loader 'app')
```